gcc/rust/backend/rust-compile-intrinsic.cc

```diff
@@ -837,6 +929,118 @@ atomic_load_handler_inner (Context *ctx, TyTy::FnType *fntype, int ordering)
   return fndecl;
 }
 
+static tree
+atomic_compare_exchange_handler_inner (Context *ctx, TyTy::FnType *fntype,
+				       tree weak_value,
+				       int success_memorder_raw,
+				       int failure_memorder_raw)
+{
+  rust_assert (fntype->get_params ().size () == 3);
+  rust_assert (fntype->get_num_substitutions () == 1);
+
+  tree lookup = NULL_TREE;
+  if (check_for_cached_intrinsic (ctx, fntype, &lookup))
+    return lookup;
+
+  auto fndecl = compile_intrinsic_function (ctx, fntype);
+
+  std::vector<Bvariable *> param_vars;
+  std::vector<tree> types;
+  compile_fn_params (ctx, fntype, fndecl, &param_vars, &types);
```

Comment:
Hmm not sure if we should provide the &types optional out variable. Since you can get the types from the Bvariable list anyway. with TREE_TYPE(param_vars.at(i)->get_decl())